azurelinuxagent/common/utils/cryptutil.py

```diff
@@ -47,33 +47,29 @@ def gen_transport_cert(self, prv_file, crt_file):
                "-out {2}").format(self.openssl_cmd, prv_file, crt_file)
         rc = shellutil.run(cmd)
         if rc != 0:
-            logger.error("Failed to create {0} and {1} certificates".format(
-                prv_file, crt_file))
+            logger.error("Failed to create {0} and {1} certificates".format(prv_file, crt_file))
 
     def get_pubkey_from_prv(self, file_name):
         if not os.path.exists(file_name):
             raise IOError(errno.ENOENT, "File not found", file_name)
         else:
-            cmd = "{0} pkey -in {1} -pubout 2>/dev/null".format(self.openssl_cmd,
-                                                                file_name)
+            cmd = "{0} pkey -in {1} -pubout 2>/dev/null".format(self.openssl_cmd, file_name)
             pub = shellutil.run_get_output(cmd)[1]
             return pub
 
     def get_pubkey_from_crt(self, file_name):
         if not os.path.exists(file_name):
             raise IOError(errno.ENOENT, "File not found", file_name)
         else:
-            cmd = "{0} x509 -in {1} -pubkey -noout".format(self.openssl_cmd,
-                                                           file_name)
+            cmd = "{0} x509 -in {1} -pubkey -noout 2>/dev/null".format(self.openssl_cmd, file_name)
             pub = shellutil.run_get_output(cmd)[1]
             return pub
 
     def get_thumbprint_from_crt(self, file_name):
         if not os.path.exists(file_name):
             raise IOError(errno.ENOENT, "File not found", file_name)
         else:
-            cmd = "{0} x509 -in {1} -fingerprint -noout".format(self.openssl_cmd,
-                                                                file_name)
+            cmd = "{0} x509 -in {1} -fingerprint -noout 2>/dev/null".format(self.openssl_cmd, file_name)
```

Comment:
but when something fails, how will we know the error?

Reply:
we should add unit tests

Reply:
Before this PR, the three methods that issued openssl commands and stored the return value were `get_pubkey_from_prv`, `get_pubkey_from_crt` and `get_thumbprint_from_crt`. Only one of them, `get_pubkey_from_prv` was redirecting the stderr to /dev/null, the other two weren't, which caused two incidents so far. My change adds the redirection of stderr to the other two commands.

None of the three functions check the return code of the commands, so we would never know if openssl failed and why. However, in the two incidents, the command wasn't failing, we were just capturing unnecessary data instead of the final return value. For example, all of these warnings would get saved to the `thumbprint` variable:
```
/usr/bin/openssl x509 -in /var/lib/waagent/{cerificate_name}.crt -fingerprint -noout
/usr/bin/openssl: /usr/local/openssl/lib/libssl.so.1.0.0: no version information available (required by /usr/bin/openssl)
/usr/bin/openssl: /usr/local/openssl/lib/libssl.so.1.0.0: no version information available (required by /usr/bin/openssl)
/usr/bin/openssl: /usr/local/openssl/lib/libssl.so.1.0.0: no version information available (required by /usr/bin/openssl)
/usr/bin/openssl: /usr/local/openssl/lib/libcrypto.so.1.0.0: no version information available (required by /usr/bin/openssl)
/usr/bin/openssl: /usr/local/openssl/lib/libcrypto.so.1.0.0: no version information available (required by /usr/bin/openssl)
/usr/bin/openssl: /usr/local/openssl/lib/libcrypto.so.1.0.0: no version information available (required by /usr/bin/openssl)
/usr/bin/openssl: /usr/local/openssl/lib/libcrypto.so.1.0.0: no version information available (required by /usr/bin/openssl)
SHA1 Fingerprint={thumbprint_value}
```
By redirecting the stderr, we ensure we only capture the output of the command itself (in this case the last line).

As for unit tests, there are unit tests already for all three methods that check the correctness of the return values, not the openssl commands itself.

Reply:
As discussed offline, changing the underlying method for invoking commands to capture stderr in case of failures. As this is done by separating stdout and stderr streams, it effectively means the return value will only consists of stdout (without stderr), which is what was broken before.